transport/grpc/outbound_test.go

```diff
@@ -200,3 +203,159 @@ func TestCallStreamWithInvalidPeer(t *testing.T) {
 		}.Error(),
 	)
 }
+
+func TestCallSvcMatchSuccess(t *testing.T) {
+	matchServer := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		grpc.UnknownServiceHandler(
+			func(srv interface{}, stream grpc.ServerStream) error {
+				ctx := stream.Context()
+				// Get transport request
+				md, ok := metadata.FromIncomingContext(ctx)
+				if md == nil || !ok {
+					return yarpcerrors.Newf(yarpcerrors.CodeInternal, "cannot get metadata from ctx: %v", ctx)
+				}
+				transportRequest, err := metadataToTransportRequest(md)
+				if err != nil {
+					return err
+				}
+
+				responseWriter := newResponseWriter()
+				defer responseWriter.Close()
+
+				// Echo accepted rpc-service in response header
+				responseWriter.AddSystemHeader(ServiceHeader, transportRequest.Service)
+
+				// Send the response attributes back and end the stream.
+				if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+					// We couldn't send the response.
+					return sendErr
+				}
+				if responseWriter.md != nil {
+					stream.SetTrailer(responseWriter.md)
+				}
+				return err
+			}),
+	)
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:15213"))
+	require.NoError(t, err)
+	go func() {
+		// TODO: do we have consistency issue here? Request comes before server begin to serve
+		_ = matchServer.Serve(listener)
+	}()
+	defer matchServer.Stop()
+
+	grpcTransport := NewTransport()
+	out := grpcTransport.NewSingleOutbound(listener.Addr().String())
+
+	require.NoError(t, grpcTransport.Start())
+	require.NoError(t, out.Start())
+	defer grpcTransport.Stop()
+	defer out.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*1000)
+	defer cancel()
+	req := &transport.Request{
+		Service:   "Service",
+		Procedure: "Hello",
+		Body:      bytes.NewReader([]byte("world")),
+	}
+	_, err = out.Call(ctx, req)
+	require.NoError(t, err)
+}
+
+func TestCallSvcMatchFailed(t *testing.T) {
+	mismatchServer := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		grpc.UnknownServiceHandler(
+			func(srv interface{}, stream grpc.ServerStream) error {
+				responseWriter := newResponseWriter()
+				defer responseWriter.Close()
+
+				// Append a wrong service name
+				responseWriter.AddSystemHeader(ServiceHeader, "ThisIsAWrongSvcName")
+
+				if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
```

Comment:
It doesn't look like we are actually writing anything to the responseWriter, is this part necessary (I'm genuinely unsure)? 

In any case, let's aim to have the minimal amount of code to get the test we want.

Reply:
Should we drop this part and just use `stream.SendHeader(..)` instead [(details)](https://sourcegraph.com/github.com/grpc/grpc-go@41344da2231b913fa3d983840a57a6b1b7b631a1/-/blob/stream.go#L626:2)?

Reply:
Sorry, this part of test codes are very verbose. I've tried using SendHeader() before, but I got "EOF" encoding error. This error disappeared when I use SendMsg(), I suspect maybe the usage of SendHeader() need to be paired with SendBody()